File: python/simpleworkflow/poller.py
"""
Workflow Run Poller for Python Workers
Polls workflow_run table and executes Python workflows
"""

import json
import logging
import os
import random
import socket
import time
from abc import ABC, abstractmethod
from datetime import datetime, timedelta
from typing import Optional, Dict, Any, Callable, List
from urllib.parse import urlparse, parse_qs, urlunparse
import psycopg2
from psycopg2.extras import RealDictCursor

from .metrics import MetricsCollector

logger = logging.getLogger(__name__)


class WorkflowRun:
    """Represents a claimed workflow run with heartbeat and cancellation support"""

    def __init__(
        self,
        run_id: str,
        workflow_type: str,
        payload: bytes,
        attempt: int,
        max_attempts: int,
        db_url: str,
        search_path: str,
        heartbeat_fn: Callable[[int], None],
        is_cancelled_fn: Callable[[], bool],
    ):
        self.id = run_id
        self.type = workflow_type
        self.payload = payload
        self.attempt = attempt
        self.max_attempts = max_attempts
        self._db_url = db_url
        self._search_path = search_path
        self._heartbeat_fn = heartbeat_fn
        self._is_cancelled_fn = is_cancelled_fn

    def heartbeat(self, duration_seconds: int = 30):
        """Extend the lease on this workflow run"""
        self._heartbeat_fn(duration_seconds)

    def is_cancelled(self) -> bool:
        """Check if this workflow run has been cancelled"""
        return self._is_cancelled_fn()


class WorkflowExecutor(ABC):
    """
    Base class for workflow executors.
    Users implement this interface to execute specific workflows.
    """

    @abstractmethod
    def execute(self, run: WorkflowRun) -> Any:
        """
        Execute a workflow run and return the result.

        Args:
            run: WorkflowRun object containing id, type, payload, attempt, max_attempts,
                 plus heartbeat() and is_cancelled() methods

        Returns:
            Any result that can be JSON-serialized

        Raises:
            Exception: If the workflow execution fails
        """
        pass


class FunctionExecutorAdapter(WorkflowExecutor):
    """Adapter that wraps a function as a WorkflowExecutor"""

    def __init__(self, func: Callable[[WorkflowRun], Any]):
        self.func = func

    def execute(self, run: WorkflowRun) -> Any:
        return self.func(run)


class IntentPoller:
    """Polls workflow_run table and executes Python workflows"""

    def __init__(
        self,
        db_url: str,
        type_prefixes: Optional[List[str]] = None,
        worker_id: Optional[str] = None,
        poll_interval: int = 2,
        lease_duration: int = 30,
        metrics: Optional[MetricsCollector] = None,
    ):
        """
        Initialize the workflow run poller with simplified API.

        Args:
            db_url: PostgreSQL connection string (may include ?schema= or ?search_path= parameter)
                    Example: "postgres://user:pass@localhost/db?schema=workflow"
            type_prefixes: List of type prefixes this poller handles (e.g., ["billing.%"])
                          If None, auto-detects from registered handlers (recommended)
            worker_id: Unique identifier for this worker
                      If None, defaults to "hostname-pid"
            poll_interval: Polling interval in seconds (default: 2)
            lease_duration: Lease duration in seconds (default: 30)
            metrics: Optional metrics collector for observability
        """
        # Parse URL to extract and handle search_path/schema
        parsed = urlparse(db_url)
        if parsed.query:
            query_params = parse_qs(parsed.query)

            # Check for schema parameter (our custom parameter)
            if 'schema' in query_params:
                self.search_path = query_params['schema'][0]
                filtered_params = {k: v for k, v in query_params.items() if k != 'schema'}
            elif 'search_path' in query_params:
                self.search_path = query_params['search_path'][0]
                filtered_params = {k: v for k, v in query_params.items() if k != 'search_path'}
            else:
                self.search_path = 'workflow'
                filtered_params = query_params

            # Rebuild query string
            new_query = '&'.join(f"{k}={v[0]}" for k, v in filtered_params.items())
            # Rebuild URL
            self.db_url = urlunparse((
                parsed.scheme, parsed.netloc, parsed.path,
                parsed.params, new_query, parsed.fragment
            ))
        else:
            self.db_url = db_url
            self.search_path = 'workflow'

        self.type_prefixes = type_prefixes
        self.auto_detect_prefix = type_prefixes is None

        # Generate default worker ID: hostname-pid
        if worker_id is None:
            hostname = socket.gethostname() or 'unknown'
            pid = os.getpid()
            self.worker_id = f"{hostname}-{pid}"
        else:
            self.worker_id = worker_id

        self.poll_interval = poll_interval
        self.lease_duration = lease_duration
        self.running = False
        self.executors: Dict[str, WorkflowExecutor] = {}
        self.function_handlers: Dict[str, Callable] = {}  # For function-based handlers
        self.metrics = metrics
        self.start_time = time.time()

        # Debug: Log metrics initialization
        if self.metrics:
            logger.info(f"DEBUG: Metrics collector initialized: {type(self.metrics).__name__}")
        else:
            logger.warning("Metrics collector is None - metrics will not be recorded")

    def register_executor(self, workflow_type: str, executor: WorkflowExecutor):
        """
        Register a workflow executor for a specific workflow type.

        Args:
            workflow_type: Workflow type (e.g., "myapp.process.v1")
            executor: WorkflowExecutor instance that handles this workflow
        """
        self.executors[workflow_type] = executor
        logger.info(f"Registered executor for workflow type: {workflow_type}")

    def handle_func(self, workflow_type: str, func: Callable[[WorkflowRun], Any]) -> 'IntentPoller':
        """
        Register a function handler for a workflow type (simplified API).
        No need to create an executor class.

        Args:
            workflow_type: Workflow type (e.g., "billing.invoice.v1")
            func: Function that takes a WorkflowRun and returns a result

        Returns:
            self for method chaining

        Example:
            poller.handle_func("billing.invoice.v1", lambda run: {...})

            # Or as decorator:
            @poller.handler("billing.invoice.v1")
            def process_invoice(run):
                return result
        """
        self.function_handlers[workflow_type] = func
        # Wrap function as executor
        self.executors[workflow_type] = FunctionExecutorAdapter(func)
        logger.info(f"Registered function handler for workflow type: {workflow_type}")
        return self

    def handler(self, workflow_type: str):
        """
        Decorator for registering workflow handlers.

        Example:
            @poller.handler("billing.invoice.v1")
            def process_invoice(run: WorkflowRun):
                payload = run.payload
                # Process invoice
                return {"status": "completed"}
        """
        def decorator(func: Callable[[WorkflowRun], Any]):
            self.handle_func(workflow_type, func)
            return func
        return decorator

    def _detect_type_prefixes(self) -> List[str]:
        """Auto-detect type prefixes from registered handlers"""
        prefix_set = set()

        for workflow_type in self.executors.keys():
            # Skip if already a prefix pattern
            if workflow_type.endswith('%'):
                prefix_set.add(workflow_type)
                continue

            # Extract prefix (everything up to first dot + %)
            parts = workflow_type.split('.')
            if len(parts) > 1:
                # Use domain prefix (e.g., "billing.invoice.v1" -> "billing.%")
                prefix = parts[0] + '.%'
                prefix_set.add(prefix)
            else:
                # No dots, use exact match
                prefix_set.add(workflow_type)

        return list(prefix_set)

    def _connect(self):
        """Create a database connection and set search_path"""
        conn = psycopg2.connect(self.db_url)
        if self.search_path:
            with conn.cursor() as cur:
                cur.execute(f"SET search_path TO {self.search_path}")
                conn.commit()
        return conn

    def start(self):
        """Start polling loop (blocking)"""
        # Auto-detect type prefixes if needed
        if self.auto_detect_prefix and not self.type_prefixes:
            self.type_prefixes = self._detect_type_prefixes()

        # Validate that at least one handler is registered
        if not self.executors:
            raise ValueError("No workflow handlers registered. Use handle_func() or register_executor() to register handlers.")

        self.running = True
        logger.info(
            f"Workflow poller started, watching type prefixes: {self.type_prefixes}"
        )
        logger.info(f"Worker ID: {self.worker_id}")

        while self.running:
            try:
                self.poll_and_execute()
            except Exception as e:
                logger.error(f"Error in poll loop: {e}", exc_info=True)

            time.sleep(self.poll_interval)

    def stop(self):
        """Stop polling"""
        self.running = False
        logger.info("Workflow poller stopped")

    def poll_and_execute(self):
        """Poll for work and execute if found"""
        # Record poll cycle and update worker metrics
        if self.metrics:
            try:
                self.metrics.record_poll_cycle(self.worker_id)
                uptime = time.time() - self.start_time
                self.metrics.update_worker_uptime(self.worker_id, uptime)
                self.metrics.update_last_poll_timestamp(self.worker_id, time.time())
            except Exception as e:
                logger.error(f"Error recording poll metrics: {e}", exc_info=True)

        execution_start = time.time()

        try:
            run = self.claim_run()
        except Exception as e:
            if self.metrics:
                try:
                    error_type = type(e).__name__
                    self.metrics.record_poll_error(self.worker_id, error_type)
                except Exception as me:
                    logger.error(f"Error recording poll error metric: {me}", exc_info=True)
            raise

        if run is None:
            # No work available - update queue depth metrics
            if self.metrics:
                try:
                    self._update_queue_depth()
                except Exception as e:
                    logger.error(f"Error updating queue depth: {e}", exc_info=True)
            return

        # Record run claimed
        logger.info(f"Claimed workflow run: {run.id} (type: {run.type})")
        if self.metrics:
            try:
                self.metrics.record_intent_claimed(run.type, self.worker_id)
                logger.info(f"DEBUG: Recorded intent_claimed metric for {run.type}")
            except Exception as e:
                logger.error(f"Error recording intent_claimed metric: {e}", exc_info=True)

        # Log started event
        self._log_event(run.id, "started", {"worker_id": self.worker_id})

        try:
            result = self.execute_run(run)
            self.mark_succeeded(run.id, run.type, result, execution_start)
        except Exception as e:
            logger.error(f"Workflow run {run.id} failed: {e}", exc_info=True)
            self.mark_failed(
                run.id,
                run.type,
                run.attempt,
                run.max_attempts,
                str(e),
                execution_start,
            )

    def claim_run(self) -> Optional[WorkflowRun]:
        """Claim a pending workflow run using atomic UPDATE...RETURNING with SKIP LOCKED"""
        conn = None
        try:
            conn = self._connect()
            with conn.cursor(cursor_factory=RealDictCursor) as cur:
                # Build parameterized type-prefix LIKE conditions
                params = [
                    self.worker_id,
                    str(self.lease_duration) + " seconds",
                ]

                type_condition = ""
                if self.type_prefixes:
                    likes = []
                    for prefix in self.type_prefixes:
                        likes.append("type LIKE %s")
                        params.append(prefix)
                    type_condition = " AND (" + " OR ".join(likes) + ")"

                # Atomic claim: UPDATE...RETURNING with SKIP LOCKED subquery
                query = f"""
                    UPDATE workflow_run
                    SET status = 'leased',
                        leased_by = %s,
                        lease_until = NOW() + (%s)::interval,
                        updated_at = NOW()
                    WHERE id = (
                        SELECT id FROM workflow_run
                        WHERE status = 'pending'
                          AND run_at <= NOW()
                          AND deleted_at IS NULL
                          {type_condition}
                        ORDER BY priority ASC, created_at ASC
                        FOR UPDATE SKIP LOCKED
                        LIMIT 1
                    )
                    RETURNING id, type, payload, attempt, max_attempts
                """
                cur.execute(query, tuple(params))

                row = cur.fetchone()
                if row is None:
                    return None

                conn.commit()

                # Create WorkflowRun object with heartbeat and cancellation functions
                run = WorkflowRun(
                    run_id=row["id"],
                    workflow_type=row["type"],
                    payload=row["payload"],
                    attempt=row["attempt"],
                    max_attempts=row["max_attempts"],
                    db_url=self.db_url,
                    search_path=self.search_path,
                    heartbeat_fn=lambda duration: self._heartbeat(row["id"], duration),
                    is_cancelled_fn=lambda: self._is_cancelled(row["id"]),
                )

                # Log leased event
                self._log_event(row["id"], "leased", {
                    "worker_id": self.worker_id,
                    "attempt": row["attempt"],
                })

                return run
        except Exception as e:
            logger.error(f"Error claiming workflow run: {e}", exc_info=True)
            if conn:
                conn.rollback()
            return None
        finally:
            if conn:
                conn.close()

    def execute_run(self, run: WorkflowRun) -> Any:
        """Execute workflow based on registered executor"""
        workflow_type = run.type

        # Find registered executor
        executor = self.executors.get(workflow_type)
        if executor is None:
            raise ValueError(
                f"No executor registered for workflow type: {workflow_type}. "
                f"Available executors: {list(self.executors.keys())}"
            )

        # Execute workflow
        logger.info(f"Executing workflow type: {workflow_type}")
        result = executor.execute(run)
        return result

    def mark_succeeded(
        self, run_id: str, workflow_type: str, result: Any, execution_start: float
    ):
        """Mark workflow run as succeeded"""
        conn = None
        try:
            conn = self._connect()
            with conn.cursor() as cur:
                cur.execute(
                    """
                    UPDATE workflow_run
                    SET status = 'succeeded',
                        result = %s,
                        updated_at = NOW()
                    WHERE id = %s
                """,
                    (json.dumps(result), run_id),
                )
                conn.commit()
                logger.info(f"Workflow run {run_id} succeeded")

                # Log succeeded event
                self._log_event(run_id, "succeeded", None)

                # Record metrics
                if self.metrics:
                    try:
                        duration = time.time() - execution_start
                        self.metrics.record_intent_completed(
                            workflow_type, self.worker_id, "succeeded", duration
                        )
                        logger.info(f"DEBUG: Recorded intent_completed metric (succeeded) for {workflow_type}, duration={duration:.3f}s")
                    except Exception as e:
                        logger.error(f"Error recording intent_completed metric: {e}", exc_info=True)
        except Exception as e:
            logger.error(
                f"Error marking workflow run {run_id} as succeeded: {e}", exc_info=True
            )
            if conn:
                conn.rollback()
        finally:
            if conn:
                conn.close()

    def mark_failed(
        self,
        run_id: str,
        workflow_type: str,
        attempt: int,
        max_attempts: int,
        error: str,
        execution_start: float,
    ):
        """Mark workflow run as failed with retry logic"""
        new_attempt = attempt + 1
        status = "pending" if new_attempt < max_attempts else "failed"

        # Exponential backoff with jitter (prevents thundering herd)
        base_delay_minutes = new_attempt ** 2
        jitter_minutes = base_delay_minutes * 0.1 * random.random()
        run_at = datetime.now() + timedelta(minutes=base_delay_minutes + jitter_minutes)

        conn = None
        try:
            conn = self._connect()
            with conn.cursor() as cur:
                cur.execute(
                    """
                    UPDATE workflow_run
                    SET status = %s,
                        attempt = %s,
                        run_at = %s,
                        last_error = %s,
                        updated_at = NOW()
                    WHERE id = %s
                """,
                    (status, new_attempt, run_at, error, run_id),
                )
                conn.commit()
                logger.info(
                    f"Workflow run {run_id} failed (attempt {new_attempt}/{max_attempts})"
                )

                # Log event
                event_type = "retried" if status == "pending" else "failed"
                self._log_event(run_id, event_type, {
                    "attempt": new_attempt,
                    "error": error,
                })

                # Record metrics
                if self.metrics:
                    try:
                        duration = time.time() - execution_start

                        # Record failed attempt
                        self.metrics.record_failed_attempt(
                            workflow_type, self.worker_id, new_attempt
                        )

                        # Record completion status
                        self.metrics.record_intent_completed(
                            workflow_type, self.worker_id, status, duration
                        )
                        logger.info(f"DEBUG: Recorded intent_completed metric ({status}) for {workflow_type}, duration={duration:.3f}s")

                        # Record failed metric if workflow permanently failed
                        if status == "failed":
                            self.metrics.record_intent_deadletter(
                                workflow_type, self.worker_id
                            )
                            logger.info(f"DEBUG: Recorded intent_deadletter metric for {workflow_type}")
                    except Exception as e:
                        logger.error(f"Error recording metrics in mark_failed: {e}", exc_info=True)
        except Exception as e:
            logger.error(
                f"Error marking workflow run {run_id} as failed: {e}", exc_info=True
            )
            if conn:
                conn.rollback()
        finally:
            if conn:
                conn.close()

    def _update_queue_depth(self):
        """Update queue depth metrics for all type prefixes"""
        if not self.metrics:
            return

        conn = None
        try:
            conn = self._connect()
            with conn.cursor() as cur:
                for prefix in self.type_prefixes:
                    cur.execute(
                        """
                        SELECT COUNT(*)
                        FROM workflow_run
                        WHERE type LIKE %s AND status = 'pending' AND deleted_at IS NULL
                    """,
                        (prefix,),
                    )
                    depth = cur.fetchone()[0]
                    self.metrics.record_queue_depth(prefix, depth)
        except Exception as e:
            logger.error(f"Error updating queue depth: {e}", exc_info=True)
        finally:
            if conn:
                conn.close()

    def _heartbeat(self, run_id: str, duration_seconds: int):
        """Extend the lease on a workflow run"""
        conn = None
        try:
            conn = self._connect()
            with conn.cursor() as cur:
                cur.execute(
                    """
                    UPDATE workflow_run
                    SET lease_until = NOW() + (%s || ' seconds')::interval,
                        updated_at = NOW()
                    WHERE id = %s AND status = 'leased'
                """,
                    (duration_seconds, run_id),
                )
                conn.commit()

                # Log heartbeat event
                self._log_event(run_id, "heartbeat", {
                    "extended_by_seconds": duration_seconds,
                })
        except Exception as e:
            logger.error(f"Error extending lease for {run_id}: {e}", exc_info=True)
            if conn:
                conn.rollback()
        finally:
            if conn:
                conn.close()

    def _is_cancelled(self, run_id: str) -> bool:
        """Check if a workflow run has been cancelled"""
        conn = None
        try:
            conn = self._connect()
            with conn.cursor() as cur:
                cur.execute(
                    "SELECT status FROM workflow_run WHERE id = %s",
                    (run_id,),
                )
                row = cur.fetchone()
                if row:
                    return row[0] == "cancelled"
                return False
        except Exception as e:
            logger.error(f"Error checking cancellation for {run_id}: {e}", exc_info=True)
            return False
        finally:
            if conn:
                conn.close()

    def _log_event(self, workflow_id: str, event_type: str, data: Optional[Dict[str, Any]]):
        """Log an audit event (best-effort, errors are ignored)"""
        conn = None
        try:
            conn = self._connect()
            with conn.cursor() as cur:
                data_json = json.dumps(data) if data else None
                cur.execute(
                    """
                    INSERT INTO workflow_event (workflow_id, event_type, data)
                    VALUES (%s, %s, %s)
                """,
                    (workflow_id, event_type, data_json),
                )
                conn.commit()
        except Exception:
            # Silently ignore event logging errors
            pass
        finally:
            if conn:
                conn.close()

File: poller.go
package simpleworkflow

import (
	"context"
	"database/sql"
	"fmt"
	"log/slog"
	"os"
	"strings"
	"sync"
	"time"
)

// PollerConfig configures the workflow poller
type PollerConfig struct {
	TypePrefixes  []string      // e.g. ["billing.%", "media.%"] for type-prefix routing
	LeaseDuration time.Duration // Lease duration (default: 30s)
	PollInterval  time.Duration // Poll interval (default: 2s)
	WorkerID      string        // Worker identifier (default: "go-worker")
}

// Poller polls workflow_run table and executes workflows (Worker API)
type Poller struct {
	db               *sql.DB
	dialect          Dialect
	runs             *RunRepository
	typePrefixes     []string // e.g. ["billing.%", "media.%"]
	executors        map[string]WorkflowExecutor
	pollInterval     time.Duration
	leaseDuration    time.Duration
	workerID         string
	stopCh           chan struct{}
	stopOnce         sync.Once
	metrics          MetricsCollector // Optional: metrics collector for observability
	startTime        time.Time        // Worker start time for uptime calculation
	autoDetectPrefix bool             // true if type prefixes should be auto-detected from handlers
	autoMigrate      bool             // true if AutoMigrate should run on Start()

	wg sync.WaitGroup // tracks in-flight executions for graceful shutdown

	// Schedule ticker (optional, enabled via WithScheduleTicker)
	scheduleTicker         *ScheduleTicker
	scheduleTickerEnabled  bool
}

// NewPoller creates a new workflow run poller from a PostgreSQL connection string.
// Type prefixes are auto-detected from registered handlers.
// Use fluent methods to configure: WithWorkerID(), WithLeaseDuration(), etc.
//
// Example:
//
//	poller, err := NewPoller("postgres://user:pass@localhost/db?schema=workflow")
//	poller.HandleFunc("billing.invoice.v1", handler)
//	poller.Start(ctx)
func NewPoller(connString string) (*Poller, error) {
	dialect, dsn, err := DetectDialect(connString)
	if err != nil {
		return nil, fmt.Errorf("failed to parse connection string: %w", err)
	}

	db, err := dialect.OpenDB(dsn)
	if err != nil {
		return nil, fmt.Errorf("failed to open database: %w", err)
	}

	if err := db.Ping(); err != nil {
		db.Close()
		return nil, fmt.Errorf("failed to connect to database: %w", err)
	}

	// Generate default worker ID
	hostname, _ := os.Hostname()
	if hostname == "" {
		hostname = "unknown"
	}
	workerID := fmt.Sprintf("%s-%d", hostname, os.Getpid())

	return &Poller{
		db:               db,
		dialect:          dialect,
		runs:             NewRunRepository(db, dialect),
		typePrefixes:     nil, // Will be auto-detected
		executors:        make(map[string]WorkflowExecutor),
		pollInterval:     2 * time.Second,  // default
		leaseDuration:    30 * time.Second, // default
		workerID:         workerID,
		stopCh:           make(chan struct{}),
		autoDetectPrefix: true,
	}, nil
}

// NewPollerWithDB creates a new workflow run poller from an existing *sql.DB and Dialect.
// Useful for testing or when you manage the connection yourself.
func NewPollerWithDB(db *sql.DB, dialect Dialect) *Poller {
	hostname, _ := os.Hostname()
	if hostname == "" {
		hostname = "unknown"
	}
	workerID := fmt.Sprintf("%s-%d", hostname, os.Getpid())

	return &Poller{
		db:               db,
		dialect:          dialect,
		runs:             NewRunRepository(db, dialect),
		typePrefixes:     nil,
		executors:        make(map[string]WorkflowExecutor),
		pollInterval:     2 * time.Second,
		leaseDuration:    30 * time.Second,
		workerID:         workerID,
		stopCh:           make(chan struct{}),
		autoDetectPrefix: true,
	}
}

// WithWorkerID sets a custom worker ID.
// Default: hostname-pid
func (p *Poller) WithWorkerID(id string) *Poller {
	p.workerID = id
	return p
}

// WithLeaseDuration sets the lease duration for claimed workflow runs.
// Default: 30 seconds
func (p *Poller) WithLeaseDuration(d time.Duration) *Poller {
	p.leaseDuration = d
	return p
}

// WithPollInterval sets how often to poll for new workflow runs.
// Default: 2 seconds
func (p *Poller) WithPollInterval(d time.Duration) *Poller {
	p.pollInterval = d
	return p
}

// WithAutoMigrate enables automatic schema migration when Start() is called.
func (p *Poller) WithAutoMigrate() *Poller {
	p.autoMigrate = true
	return p
}

// WithScheduleTicker enables the schedule ticker inside the poller.
// The ticker converts due workflow_schedule rows into workflow_run rows.
func (p *Poller) WithScheduleTicker() *Poller {
	p.scheduleTickerEnabled = true
	return p
}

// WithScheduleTickInterval sets the tick interval for the embedded schedule ticker.
// Default: 15 seconds. Only effective if WithScheduleTicker() is also called.
func (p *Poller) WithScheduleTickInterval(d time.Duration) *Poller {
	p.scheduleTickerEnabled = true
	if p.scheduleTicker == nil {
		p.scheduleTicker = newScheduleTickerFromDB(p.db, p.dialect)
	}
	p.scheduleTicker.tickInterval = d
	return p
}

// WithTypePrefixes explicitly sets type prefixes to watch.
// This overrides auto-detection from registered handlers.
// Example: WithTypePrefixes("billing.%", "notify.%")
func (p *Poller) WithTypePrefixes(prefixes ...string) *Poller {
	p.typePrefixes = prefixes
	p.autoDetectPrefix = false
	return p
}

// HandleFunc registers a function handler for a workflow type.
// The function receives the WorkflowRun and returns a result or error.
//
// Example:
//
//	poller.HandleFunc("billing.invoice.v1", func(ctx context.Context, run *WorkflowRun) (any, error) {
//	    // Process invoice
//	    return result, nil
//	})
func (p *Poller) HandleFunc(workflowType string, fn func(context.Context, *WorkflowRun) (any, error)) *Poller {
	p.executors[workflowType] = &funcExecutorAdapter{fn: fn}
	return p
}

// Handle registers a WorkflowExecutor for a workflow type.
// Use this for advanced cases where you need a stateful executor.
func (p *Poller) Handle(workflowType string, executor WorkflowExecutor) *Poller {
	p.executors[workflowType] = executor
	return p
}

// Close closes the database connection.
func (p *Poller) Close() error {
	if p.db != nil {
		return p.db.Close()
	}
	return nil
}

// funcExecutorAdapter adapts a simple function to the WorkflowExecutor interface.
type funcExecutorAdapter struct {
	fn func(context.Context, *WorkflowRun) (any, error)
}

func (a *funcExecutorAdapter) Execute(ctx context.Context, run *WorkflowRun) (any, error) {
	return a.fn(ctx, run)
}

// SetMetrics sets the metrics collector (optional, pass nil to disable metrics)
func (p *Poller) SetMetrics(m MetricsCollector) {
	p.metrics = m
	p.startTime = time.Now()
}

// Start begins polling for workflow runs
func (p *Poller) Start(ctx context.Context) {
	// Run auto-migration if enabled
	if p.autoMigrate {
		if err := p.AutoMigrate(ctx); err != nil {
			slog.Error("auto-migrate failed", "error", err)
			os.Exit(1)
		}
	}

	// Auto-detect type prefixes from registered handlers if needed
	if p.autoDetectPrefix && len(p.typePrefixes) == 0 {
		p.typePrefixes = p.detectTypePrefixes()
	}

	// Validate that at least one handler is registered
	if len(p.executors) == 0 {
		slog.Error("no workflow handlers registered, use Handle() or HandleFunc() to register handlers")
		os.Exit(1)
	}

	// Start schedule ticker goroutine if enabled
	if p.scheduleTickerEnabled {
		if p.scheduleTicker == nil {
			p.scheduleTicker = newScheduleTickerFromDB(p.db, p.dialect)
		}
		if p.metrics != nil {
			p.scheduleTicker.SetMetrics(p.metrics)
		}
		go p.scheduleTicker.Start(ctx)
		slog.Info("embedded schedule ticker started", "interval", p.scheduleTicker.tickInterval)
	}

	ticker := time.NewTicker(p.pollInterval)
	defer ticker.Stop()

	slog.Info("workflow poller started", "type_prefixes", p.typePrefixes)

	for {
		select {
		case <-ticker.C:
			p.pollAndExecute(ctx)
		case <-p.stopCh:
			if p.scheduleTicker != nil {
				p.scheduleTicker.Stop()
			}
			p.wg.Wait()
			return
		case <-ctx.Done():
			if p.scheduleTicker != nil {
				p.scheduleTicker.Stop()
			}
			p.wg.Wait()
			return
		}
	}
}

// detectTypePrefixes extracts type prefixes from registered handlers.
// It auto-detects common prefixes to minimize polling overhead.
func (p *Poller) detectTypePrefixes() []string {
	prefixMap := make(map[string]bool)

	for workflowType := range p.executors {
		// Skip if it's already a prefix pattern
		if strings.HasSuffix(workflowType, "%") {
			prefixMap[workflowType] = true
			continue
		}

		// Extract prefix (everything up to last dot + %)
		parts := strings.Split(workflowType, ".")
		if len(parts) > 1 {
			// Use domain prefix (e.g., "billing.invoice.v1" -> "billing.%")
			prefix := parts[0] + ".%"
			prefixMap[prefix] = true
		} else {
			// No dots, use exact match
			prefixMap[workflowType] = true
		}
	}

	// Convert map to slice
	prefixes := make([]string, 0, len(prefixMap))
	for prefix := range prefixMap {
		prefixes = append(prefixes, prefix)
	}

	return prefixes
}

// Stop stops the poller. Safe to call multiple times.
func (p *Poller) Stop() {
	p.stopOnce.Do(func() { close(p.stopCh) })
}

func (p *Poller) pollAndExecute(ctx context.Context) {
	// Record poll cycle (automatically updates uptime and last poll timestamp)
	if p.metrics != nil {
		p.metrics.RecordPollCycle(p.workerID)
	}

	executionStart := time.Now()

	// Claim a workflow run
	run, err := p.claimRun(ctx)
	if err != nil {
		if p.metrics != nil {
			p.metrics.RecordPollError(p.workerID, classifyError(err))
		}
		slog.Error("failed to claim workflow run", "error", err)
		return
	}
	if run == nil {
		// No work available - update queue depth metrics
		if p.metrics != nil {
			p.updateQueueDepth(ctx)
		}
		return
	}

	// Record run claim
	if p.metrics != nil {
		p.metrics.RecordIntentClaimed(run.Type, p.workerID)
	}

	slog.Info("claimed workflow run", "run_id", run.ID, "type", run.Type)

	// Execute the workflow
	p.wg.Add(1)
	defer p.wg.Done()
	p.executeRun(ctx, run, executionStart)
}

func (p *Poller) claimRun(ctx context.Context) (*WorkflowRun, error) {
	run, err := p.runs.Claim(ctx, p.workerID, p.typePrefixes, p.leaseDuration)
	if err != nil || run == nil {
		return run, err
	}

	// Attach heartbeat and cancellation check functions
	run.Heartbeat = p.makeHeartbeatFunc(run.ID)
	run.IsCancelled = p.makeCancellationCheckFunc(run.ID)

	return run, nil
}

func (p *Poller) executeRun(ctx context.Context, run *WorkflowRun, executionStart time.Time) {
	// Find executor for this workflow type
	executor, ok := p.executors[run.Type]
	if !ok {
		err := fmt.Errorf("no executor registered for workflow type: %s", run.Type)
		p.markRunFailed(ctx, run, err, executionStart)
		return
	}

	// Log started event (best-effort)
	p.runs.logEvent(ctx, run.ID, "started", map[string]any{
		"worker_id": p.workerID,
	})

	// Execute workflow
	result, err := executor.Execute(ctx, run)

	// Update run status
	if err != nil {
		p.markRunFailed(ctx, run, err, executionStart)
	} else {
		p.markRunSucceeded(ctx, run, result, executionStart)
	}
}

func (p *Poller) markRunSucceeded(ctx context.Context, run *WorkflowRun, result any, executionStart time.Time) {
	if err := p.runs.MarkSucceeded(ctx, run.ID, result); err != nil {
		slog.Error("failed to mark workflow run as succeeded", "run_id", run.ID, "error", err)
	} else {
		slog.Info("workflow run succeeded", "run_id", run.ID)
	}

	// Record metrics
	if p.metrics != nil {
		duration := time.Since(executionStart)
		p.metrics.RecordIntentCompleted(run.Type, p.workerID, "succeeded", duration)
	}
}

func (p *Poller) markRunFailed(ctx context.Context, run *WorkflowRun, execErr error, executionStart time.Time) {
	status := p.runs.MarkFailed(ctx, run, execErr)

	// Record metrics
	if p.metrics != nil {
		duration := time.Since(executionStart)
		newAttempt := run.Attempt + 1

		p.metrics.RecordFailedAttempt(run.Type, p.workerID, newAttempt)
		p.metrics.RecordIntentCompleted(run.Type, p.workerID, status, duration)

		if status == "failed" {
			p.metrics.RecordIntentDeadletter(run.Type, p.workerID)
		}
	}
}

// updateQueueDepth queries and updates queue depth metrics for all type prefixes
func (p *Poller) updateQueueDepth(ctx context.Context) {
	if p.metrics == nil {
		return
	}

	for _, prefix := range p.typePrefixes {
		depth, err := p.runs.CountPending(ctx, prefix)
		if err == nil {
			p.metrics.RecordQueueDepth(prefix, "pending", depth)
		}
	}
}

// makeHeartbeatFunc creates a heartbeat function for extending workflow run lease
func (p *Poller) makeHeartbeatFunc(runID string) HeartbeatFunc {
	return func(ctx context.Context, duration time.Duration) error {
		return p.runs.ExtendLease(ctx, runID, duration)
	}
}

// makeCancellationCheckFunc creates a function to check if workflow run has been cancelled
func (p *Poller) makeCancellationCheckFunc(runID string) CancellationCheckFunc {
	return func(ctx context.Context) (bool, error) {
		return p.runs.CheckCancelled(ctx, runID)
	}
}

// classifyError categorizes errors for metrics tracking
func classifyError(err error) string {
	if err == nil {
		return "none"
	}
	switch err {
	case sql.ErrNoRows:
		return "no_rows"
	case sql.ErrTxDone:
		return "tx_done"
	case context.DeadlineExceeded:
		return "timeout"
	case context.Canceled:
		return "canceled"
	default:
		if err.Error() == "database is closed" {
			return "db_closed"
		}
		return "unknown"
	}
}
